addons/mail/controllers/discuss.py

```diff
@@ -406,16 +406,30 @@ def mail_channel_add_guest_as_member(self, channel_id, channel_uuid, **kwargs):
                 raise NotFound()
 
     @http.route('/mail/channel/messages', methods=['POST'], type='json', auth='public')
-    def mail_channel_messages(self, channel_id, max_id=None, min_id=None, limit=30, **kwargs):
+    def mail_channel_messages(self, channel_id, max_id=None, min_id=None, limit=30, order=None, **kwargs):
         channel_member_sudo = request.env['mail.channel.member']._get_as_sudo_from_request_or_raise(request=request, channel_id=int(channel_id))
-        messages = channel_member_sudo.env['mail.message']._message_fetch(domain=[
+        domain = [
             ('res_id', '=', channel_id),
             ('model', '=', 'mail.channel'),
             ('message_type', '!=', 'user_notification'),
-        ], max_id=max_id, min_id=min_id, limit=limit)
-        if not request.env.user._is_public():
+        ]
+        if 'around' in kwargs:
+            messages_before = channel_member_sudo.env['mail.message']._message_fetch(domain=[*domain, ('id', '<=', kwargs['around'])], limit=limit // 2)
+            messages_after = channel_member_sudo.env['mail.message']._message_fetch(domain=[*domain, ('id', '>', kwargs['around'])], limit=limit // 2, order='id')
+            messages = (messages_after + messages_before)
+        else:
+            messages = channel_member_sudo.env['mail.message']._message_fetch(domain=domain, max_id=max_id, min_id=min_id, limit=limit, order=order)
+        if not request.env.user._is_public() and 'around' not in kwargs:
             messages.set_message_done()
-        return messages.message_format()
+        return messages.sorted('id', reverse=True).message_format()
+
+    @http.route('/mail/channel/pins', methods=['POST'], type='json', auth='public')
+    def mail_channel_pins(self, channel_id, max_id=None, **kwargs):
+        channel_member_sudo = request.env['mail.channel.member']._get_as_sudo_from_request_or_raise(request=request, channel_id=int(channel_id))
+        pinned_message_ids = channel_member_sudo.channel_id.pinned_message_ids
+        if max_id:
```

Comment:
why a max for pin?

Reply:
We do not want to fetched already known messages (the one the thread has already loaded in its message list)

Reply:
It's not a big deal to fetch the same message again.
It's best to keep pinned code simple and just have it fetch all fetched messages.